day-03/part-2/lelithium.zig

```diff
@@ -0,0 +1,185 @@
+
+    for (input) |char| {
+        // std.debug.print("char: {c}\n", .{char});
+        // Check if we are at the end of a line
+        if (bit_pos == LENGTH) {
+            bit_pos = 0; // reset bit_pos
+            line_count += 1; // count new line
+            continue; // skip over tests
+        }
+        // Build the first bit count (for all valid lines) so we don't have to re-walk the entire array
+        if (bit_pos == 0 and char == '1') {
+            bit_count += 1; //
+        }
+        // Store the line
+        all_lines[line_count][bit_pos] = char;
+        bit_pos += 1;
+    }
+    line_count += 1; // don't forget the final line
+
+    var full_line_count = line_count; // store full line count to rebuild valid_line_idx afterwards
+    var full_bit_count = bit_count; // also store bit count for first iteration
+
+    // Stores if the current valid value for a bit is 1
+    var bit_choice: bool = false;
+
+    // Stores both target values indexes
+    var oxygen_idx: u16 = 0;
+    var co2_idx: u16 = 0;
+
+    // Used to unroll the processing loop.
+    // Iteration 0 sets o2, and iteration 1 sets co2
+    comptime var iterations: u2 = 0;
+    inline while (iterations < 2) : (iterations += 1) {
+        // (re)build valid_line_idx
+        var _i: u16 = 0;
+        while (_i < full_line_count) : (_i += 1) {
+            valid_line_idx[_i] = _i;
+        }
+        // Reset bit_pos
+        bit_pos = 0;
+        // Reset bit count to first version
+        bit_count = full_bit_count;
+        // Reset line count
+        line_count = full_line_count;
+        main: while (bit_pos < LENGTH) : (bit_pos += 1) {
+            //std.debug.print("{} bits for {} lines\n", .{bit_count, line_count});
+            bit_choice = bit_count * 2 >= line_count;
+            if (iterations == 1) {
+                // Flip bit_choice to select least bit.
+                bit_choice = !bit_choice;
+            }
+            //std.debug.print("Now working on bit_pos {} (choice {})\n", .{bit_pos, bit_choice});
+            // Reset bit count (look-ahead)
+            bit_count = 0;
+            for (valid_line_idx) |idx| {
+                // Check if we're on a valid index
+                if (idx == 1001) {
+                    continue; // skip on 1001
+                }
+                if (idx == 1002) {
+                    break; // end on 1002
+                }
+                // If there's only one valid line left, pick it
+                if (line_count == 1) {
+                    if (iterations == 0) {
+                        oxygen_idx = idx;
+                    } else {
+                        co2_idx = idx;
+                    }
+                    // Break out of the main loop
+                    break :main;
+                }
+                //std.debug.print("\tUsing index {}\n", .{idx});
+                //std.debug.print("\tProcessing line {s}\n", .{all_lines[idx]});
+                if (bit_choice) {
```

Comment:
2 ifs imbriqués qui font la même chose avec l'un inversé, ça se simplifie avec un XOR

```rust
if a {
  if b {
    F()
  } else {
    G()
  }
} else {
  if !b {
    F()
  } else {
    G()
  }
}
// est équivalent à
if a ^ b {
  G()
} else {
  F()
}
```

Reply:
Keur sur toi

Reply:
Fixed in 5541c86